Source/Particles/ParticleCreation/FilterCopyTransform.H

```diff
@@ -133,7 +137,154 @@ Index filterCopyTransformParticles (DstTile& dst, SrcTile& src, Index dst_index,
         p_mask[i] = filter(src_data, i);
     });
 
-    return filterCopyTransformParticles(dst, src, mask.dataPtr(), dst_index,
+    return filterCopyTransformParticles<N>(dst, src, mask.dataPtr(), dst_index,
+                                                      std::forward<CopyFunc>(copy),
+                                                      std::forward<TransFunc>(transform));
+}
+
+/**
+ * \brief Apply a filter, copy, and transform operation to the particles
+ * in src, in that order, writing the results to dst1 and dst2, starting
+ * at dst1_index and dst2_index. The dst tiles will be extended so all the
+ * particles will fit, if needed.
+ *
+ * Note that the transform function operates on all of src, dst1, and dst2,
+ * so all of them can be modified.
+ *
+ * \tparam N number of particles created in the dst(s) for each filtered src particle
+ * \tparam DstTile the dst particle tile type
+ * \tparam SrcTile the src particle tile type
+ * \tparam Index the index type, e.g. unsigned int
+ * \tparam TransFunc the transform function type
+ * \tparam CopyFunc the copy function type
+ *
+ * \param dst1 the destination tile
+ * \param dst2 the destination tile
+ * \param src the source tile
+ * \param mask pointer to the mask - 1 means copy, 0 means don't copy
+ * \param dst1_index the location at which to starting writing the result to dst1
+ * \param dst2_index the location at which to starting writing the result to dst2
+ * \param copy callable that defines what will be done for the "copy" step.
+ * \param transform callable that defines the transformation to apply on dst and src.
+ *
+ *        The form of the callable should model:
+ *            template <typename DstData, typename SrcData>
+ *            AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+ *            void operator() (DstData& dst, SrcData& src, int i_src, int i_dst);
+ *
+ *        where dst and src refer to the destination and source tiles and
+ *        i_src and i_dst and the particle indices in each tile.
+ *
+ * \return num_added the number of particles that were written to dst.
+ */
+template <int N, typename DstTile, typename SrcTile, typename Index,
+          typename TransFunc, typename CopyFunc,
+          amrex::EnableIf_t<std::is_integral<Index>::value, int> foo = 0>
+Index filterCopyTransformParticles (DstTile& dst1, DstTile& dst2, SrcTile& src, Index* mask,
```

Comment:
Could this be `Index const * const mask,`?

Reply:
This seems to interfere with template overload resolution in a way I'm not sure I 100% understand. I think that both `int` and `const int`, for example, could be the `Index` type.